core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -577,14 +582,42 @@ private SnapshotInfo inProgressSnapshot(SnapshotsInProgress.Entry entry) {
                         shardSnapshotStatus.updateStage(IndexShardSnapshotStatus.Stage.FAILURE);
                         shardSnapshotStatus.failure(shardFailure.reason());
                         shardStatus.put(shardId, shardSnapshotStatus);
+                        hasFailedShards = true;
                     } else {
-                        IndexShardSnapshotStatus shardSnapshotStatus =
-                            repository.getShardSnapshotStatus(snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
-                        shardStatus.put(shardId, shardSnapshotStatus);
+                        nonFailedShards.put(shardId, indexId);
                     }
                 }
             }
         }
+        for (Map.Entry<ShardId, IndexId> shard : nonFailedShards.entrySet()) {
+            final ShardId shardId = shard.getKey();
+            final IndexId indexId = shard.getValue();
+            IndexShardSnapshotStatus shardSnapshotStatus;
+            try {
+                shardSnapshotStatus = repository.getShardSnapshotStatus(
+                    snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
+            } catch (IndexShardRestoreFailedException e) {
```

Comment:
@dadoonet the problem here is that I need to depend on the snapshot status itself now (see https://github.com/elastic/elasticsearch/pull/23833/commits/45c994d98b955c834aa8533de21e54d3798199c5) to determine if the shard status should be read, and that snapshot status is not passed in to this method.  Otherwise, it would've been a good proposition!